File: scripts/cost_logger.py
"""
cost_logger.py - Shared cost logging module for direct API calls.

Import this in any Python script that calls an AI API directly:

    import sys, os
    sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
    # or: sys.path.insert(0, '/home/openclaw/.openclaw/workspace/scripts')
    from cost_logger import log_cost

    log_cost(
        model="grok-4-1-fast-non-reasoning",
        input_tokens=1234,
        output_tokens=56,
        source="xread.py"
    )
"""

import json
import os
from datetime import datetime, timezone

LOG_FILE = "/home/openclaw/logs/direct-api-costs.jsonl"


def log_cost(model: str, input_tokens: int, output_tokens: int, source: str, timestamp_utc: str = None):
    """
    Append a cost record to the direct API costs log.

    Args:
        model:          Model name used (e.g. "grok-4-1-fast-non-reasoning")
        input_tokens:   Number of input/prompt tokens
        output_tokens:  Number of output/completion tokens
        source:         Calling script name (e.g. "xread.py")
        timestamp_utc:  ISO8601 UTC timestamp (defaults to now)
    """
    try:
        ts = timestamp_utc or datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
        record = {
            "ts": ts,
            "model": str(model),
            "input_tokens": int(input_tokens or 0),
            "output_tokens": int(output_tokens or 0),
            "source": str(source),
        }

        # Ensure log directory exists
        os.makedirs(os.path.dirname(LOG_FILE), exist_ok=True)

        # Append mode is atomic on Linux for small writes — no extra locking needed
        with open(LOG_FILE, "a") as f:
            f.write(json.dumps(record) + "\n")
    except Exception:
        # Silent failure — never crash the calling script
        pass

File: fathom/cost-logger.js
/**
 * cost-logger.js - Shared cost logging module for direct API calls.
 *
 * Usage:
 *   const { logCost } = require('./cost-logger');
 *   logCost('claude-sonnet-4-6', 1234, 56, 'fathom/processor.js');
 */

'use strict';

const fs = require('fs');
const path = require('path');

const LOG_FILE = '/home/openclaw/logs/direct-api-costs.jsonl';

/**
 * Append a cost record to the shared direct-API cost log.
 * Silent on any error — never crashes the calling script.
 *
 * @param {string} model        - Model name (e.g. "claude-sonnet-4-6-20251115")
 * @param {number} inputTokens  - Input/prompt token count
 * @param {number} outputTokens - Output/completion token count
 * @param {string} source       - Calling script name (e.g. "fathom/processor.js")
 */
function logCost(model, inputTokens, outputTokens, source) {
  try {
    const ts = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
    const record = {
      ts,
      model: String(model || ''),
      input_tokens: inputTokens || 0,
      output_tokens: outputTokens || 0,
      source: String(source || ''),
    };

    // Ensure log directory exists
    const dir = path.dirname(LOG_FILE);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    // appendFileSync is atomic on Linux for small writes
    fs.appendFileSync(LOG_FILE, JSON.stringify(record) + '\n');
  } catch (_) {
    // Silent failure — never crash the calling script
  }
}

module.exports = { logCost };
